erigon-lib/state/history.go

```diff
@@ -1062,26 +1062,33 @@ func (hc *HistoryContext) statelessIdxReader(i int) *recsplit.IndexReader {
 	return r
 }
 
-func (hc *HistoryContext) CanPruneUntil(tx kv.Tx, untilTxNum uint64) bool {
-	inSnapsTx := hc.maxTxNumInFiles(false)
+func (hc *HistoryContext) canPruneUntil(tx kv.Tx, untilTxNum uint64) (can bool, txTo uint64) {
 	minIdxTx := hc.ic.CanPruneFrom(tx)
 	maxIdxTx := hc.ic.highestTxNum(tx)
+	//defer func() {
+	//	fmt.Printf("CanPrune[%s]Until(%d) noFiles=%t txTo %d idxTx [%d-%d] keepTxInDB=%d; result %t\n", hc.h.filenameBase, untilTxNum, hc.h.dontProduceFiles, txTo, minIdxTx, maxIdxTx, hc.h.keepTxInDB, minIdxTx < txTo)
+	//}()
 
-	// if we don't produce files, we can prune only if:
-	isNoFilesAndEnoughTxKeptInDB := hc.h.dontProduceFiles && // files are not produced
-		minIdxTx != math.MaxUint64 && // idx has data
-		minIdxTx < untilTxNum && // idx data < untilTxNum
-		hc.h.keepTxInDB < maxIdxTx && // sub overflow
-		minIdxTx < maxIdxTx-hc.h.keepTxInDB // idx data < MaxTx-keepTxInDB
+	if hc.h.dontProduceFiles {
+		if hc.h.keepTxInDB >= maxIdxTx {
+			return false, 0
+		}
+		// we only could prune up until txTo
+		txTo = maxIdxTx - hc.h.keepTxInDB
+	} else {
+		txTo = hc.maxTxNumInFiles(false)
```

Comment:
Then history will have `dontProduceFiles=true` which will margin from `maxIdxTx - keepTxInDB`. 
And we need to keep pruning of history and domain separated (watch only on related files during pruning) because in some cases there is some history but no domain to prune or vice versa but in both cases we need to perform bounded prune to avoid exec problems.